Sticky scrolling in Eclipse

Christopher-Hermann opened the pull request:
**Description**
With this change, sticky scrolling is introduced into Eclipse text editors, as it is available for many other modern IDEs. Sticky scrolling will keep certain source code lines visible and in a fixed position on the screen as the user scrolls down the page. This technique improves user experience by keeping information within reach at all times.

The feature can be enabled via the TextEditor settings.

Provides feature for https://github.com/eclipse-platform/eclipse.platform.ui/issues/1719 and https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1364

**Current state**
There are some TODOs left, however the general functionality should work. I'm already creating the PR to get early feedback and maybe there are some volunteer beta testers.

**Testing**
Enable the feature in the preferences: General > Editors > Text Editors 
Check "Enable sticky scrolling" 

The sticky lines should be visible in all source/text editors and should work for every language since it is working on line indentation.
![stickyScrolling](https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/712d00b2-9baf-4798-9df0-b47d26212030)

Christopher-Hermann committed:
Provide sticky scrolling in source editors #1719

With this change, sticky scrolling is introduced into Eclipse text editors. Sticky scrolling will keep certain source code lines visible and in a fixed position on the screen as the user scrolls down the page. This technique improves user experience by keeping information within reach at all times.

The feature can be enabled via the TextEditor settings.

Provides feature for https://github.com/eclipse-platform/eclipse.platform.ui/issues/1719 and https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1364

BeckerWdf commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/texteditor/AbstractDecoratedTextEditor.java:
a lot of whitespace changes mix with "real" changes.
Can you separate the whitespace changes into a separate commit?

BeckerWdf commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/texteditor/AbstractDecoratedTextEditorPreferenceConstants.java:
Same whitespace change issues here.

AObuchow commented:
Just wanted to say this looks _awesome_ & a great modernization effort for the Eclipse IDE :) Great work

mickaelistria commented:
That's a very interesting proposal and the quality already seems good!
What defines which lines are to remain visible? Identation only?

iloveeclipse commented:
> What defines which lines are to remain visible? Identation only?

This seem to produce unexpected effects with code formatted like this:
```
    private void writeObject(java.io.ObjectOutputStream s)
        throws java.io.IOException
    {
```
see
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/964108/765182e1-efd5-492b-8b77-8adffbacd496)

iloveeclipse commented:
In general I believe adding first type in the file wastes space / just duplicates already shown information, as the first type is already shown in the editor tab title.

On Linux I see the right side of the "sticky" widget is shown one pixel over the scrollbar, as seen in the screenshot above.

Beside this, I would wish a bit different background for the sticky part, as it is hard to differentiate from actual code (also seen in the screenshot above).

iloveeclipse commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/texteditor/AbstractDecoratedTextEditor.java:
```diff
@@ -251,9 +255,11 @@ public void gotoMarker(IMarker marker) {
 	 * The overview ruler of this editor.
 	 *
-	 * <p>This field should not be referenced by subclasses. It is <code>protected</code> for API
+	 * <p>
```
There is a lot of unrelated javadoc format changes. Please undo, it is hard to find code for a review

Christopher-Hermann replied:
Done

iloveeclipse commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/texteditor/AbstractDecoratedTextEditor.java:
```diff
@@ -330,6 +344,12 @@ public void gotoMarker(IMarker marker) {
+	/**
+	 * The handler for sticky scrolling
+	 * 
+	 * @since 4.3 //TODO ?
```
So not I've found what I was looking for. This must be bundle major.minor version, which brings us to the next issue: there is new API and that means, bundle version must be updated from `3.17` to `3.18` and this is what should be added here instead of `4.3`.

But the **private** fields are never APi, so at this place it is wrong. It should be at `AbstractDecoratedTextEditorPreferenceConstants.EDITOR_STICKY_SCROLLING_ENABLED`

Christopher-Hermann replied:
Done

BeckerWdf commented:
> That's a very interesting proposal and the quality already seems good! What defines which lines are to remain visible? Identation only?

I think yes. How do other IDEs solve this?

Christopher-Hermann commented:
> > What defines which lines are to remain visible? Identation only?
> 
> This seem to produce unexpected effects with code formatted like this:
> 
> ```
>     private void writeObject(java.io.ObjectOutputStream s)
>         throws java.io.IOException
>     {
> ```

> > That's a very interesting proposal and the quality already seems good! What defines which lines are to remain visible? Identation only?
> 
> I think yes. How do other IDEs solve this?

Yes, currently it is working with indentation only. At least VS Code does the exact same thing, see the EnumMap example: 
<img width="489" alt="Bildschirmfoto 2024-05-23 um 08 21 46" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/6fcc8dda-486f-4eef-bc0a-919171d0fc76">

But of course we can try to find a better solution. 
On the first look I thought about using the code folding information. But this has some drawbacks as for example Java doesn't provide information for if else statements. Furthermore, this won't work for language that doesn't provide code folding at all.

mickaelistria commented:
> On the first look I thought about using the code folding information

That's a good idea to get started from.
As we could already notice, indentation is sometimes not sufficient. While folding may not be perfect, its internals are a good inspiration: it does rely on document annotations (which are set by JDT when reconciling/parsing the Java file). A similar strategy allowing document annotations would be interesting here, because the problem "what to stick in" cannot be fully fixed with plain text logic, it seems language-specific and even code-style-specific.
So one possible approach could be to define a new "sticky" document annotation type in Platform and show in the sticky area the lines that have this sticky annotation. Then, language analyzers such as JDT can mark some lines as "sticky" according to language-specific logic. If no line is marked as sticky, indentation can be used as fall-back, with its limits.
What do you think?

Christopher-Hermann commented:
> So one possible approach could be to define a new "sticky" document annotation type in Platform and show in the sticky area the lines that have this sticky annotation. Then, language analyzers such as JDT can mark some lines as "sticky" according to language-specific logic. If no line is marked as sticky, indentation can be used as fall-back, with its limits. What do you think?

Yes, that's a good idea. I will finish the work on the layout and on the fallback logic based on indentation. Afterwards we can have a look how we can implement the annotation logic.

Madjosz commented:
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/28844868/e6be817b-2768-4dba-b31e-4677e2cf4afd)
Eclipse Theia seems to work better in this regard then than VS code as it sticks the correct lines with class and function names in 1TBS as well as in Allman style. The sticking of the `if` brace seems to be an artifact and only happens after restart. @Christopher-Hermann are you sure this behavior keeps appearing after the file is fully loaded and parsed? I see the same behaviour in Theia during loading of the file.

Regarding redundant `class` name stick one could argue that you can have multiple non-public top level classes where you want the stick.

Having sticky scroll use the folding functionality sounds like a good approach. Having conditionals not stick yet sounds more like a missing feature on the folding side to me.

Wittmaxi commented:
This is a very important feature that eclipse is lacking (and I get asked about a lot!).
I am very glad that somebody is working on it - I am willing to test it, if you need help with that :)

Christopher-Hermann commented:
> On Linux I see the right side of the "sticky" widget is shown one pixel over the scrollbar, as seen in the screenshot above.

It seems that the layout of the scroll bar is different on the different operating systems. It would be nice to get a screenshot from a Windows user as well, so that I can address the issue.

On MacOS the scroll bar is transparent with the latest client:
![ScrollBar](https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/57469c52-ea18-432a-9b15-6f9ce74b2cd2)

> Beside this, I would wish a bit different background for the sticky part, as it is hard to differentiate from actual code (also seen in the screenshot above).

Yes you are right, it's hard to see. Any color suggestions? I would like to reuse a existing color, but I cannot find a matching one.
<img width="434" alt="Bildschirmfoto 2024-05-28 um 13 04 18" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/5f4e959f-b101-4d06-8cbd-cff642a4b817">

Christopher-Hermann commented:
> Having sticky scroll use the folding functionality sounds like a good approach. Having conditionals not stick yet sounds more like a missing feature on the folding side to me.

Yes, most probably you are right and this could also be improved in the code folding in general.

However, I propose merging this change as a "beta feature" after addressing layout questions and providing unit tests. Since the feature is disabled by default it should not have any impacts on the daily work. And having it in the official eclipse release will increase the amount of tester significantly. Even I don't use my target IDE in my daily work, so it's hard to test every single editor feature, like quick fixes, etc.

The sticky scrolling based on annotations/code folding can then be implemented on a separate PR.

Madjosz commented:
> It would be nice to get a screenshot from a Windows user as well, so that I can address the issue.

This is how it currently looks in Windows 11 light theme without OS scaling and dark theme with 125% OS scaling:
![Sticky Scroll in Windows 11](https://github.com/eclipse-platform/eclipse.platform.ui/assets/28844868/7758eb72-bd52-4c23-b0c4-997f6b1571fd)
The line number and code alignment is pixel perfect. What I don't like is the yellow background, that is too much.

Overall a very nice outcome, thank you for your efforts!

Christopher-Hermann commented:
> What I don't like is the yellow background, that is too much.

Yes, I played around a litte bit with the colors, but I don't like this as well. I removed the coloring and incorporated a second vertical separator beneath the sticky lines, which improved visibility slightly. However, there is still room for improvement.
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/166f071a-eb56-44d9-9647-419f70a0713b)


An other option would be to color the line numbers:
<img width="611" alt="Bildschirmfoto 2024-05-29 um 16 19 58" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/37f85a74-a80d-4980-a955-a6d2049a3d4c">

Christopher-Hermann commented:
The feature/pull request is now ready for review. The sticky lines calculation is fully reliant on the indentation of the source code. I propose we merge it as a "beta" feature to gather preliminary feedback, particularly from users coding in languages other than Java, which was my exclusive testing focus.

For future enhancements, the sticky scrolling functionality can be refined by introducing a sticky lines provider capable of interpreting code folding or working from a new annotation as suggested in a comment above. To allow for broader use across different languages, we could implement this with extension points.

The "final" state of the UI now contains two horizontal separators to make the separation between the sticky lines and the actual source code easier to see. Ideas for improvement are welcome:
<img width="601" alt="Bildschirmfoto 2024-06-10 um 13 06 44" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/4dfdf594-b70b-425d-82d7-9f7147c47740">

BeckerWdf commented:
> I propose we merge it as a "beta" feature to gather preliminary feedback, particularly from users coding in languages other than Java, which was my exclusive testing focus.

That's a good idea.

mickaelistria reviewed (changes requested):
I'm all in favor of merging ASAP, this is an interesting, opt-in feature with only internal code and just a couple of new symbols in API, and we're in M1 which is the best time for such addition without adding major risk on the release.
Please just add the copyright headers to new file before we merge.

iloveeclipse commented on tests/org.eclipse.ui.editors.tests/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyLinesProviderTest.java:
```diff
@@ -0,0 +1,217 @@
+package org.eclipse.ui.internal.texteditor.stickyscroll;
```
Please add header comments to all new files

Wittmaxi commented:
As said above, I really like this feature and I absolutely want to use it!
I have tested your feature and will just dump all of my findings - no matter how big or little.
To me, 1), 2) and 3) are (almost) blockers, the rest are optional but should be taken care of (maybe as a follow-up).

1) Have you looked into this change in dark mode?
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/8e06ea66-5f9a-420a-a601-1dc6d39742b6)

The double line in dark mode is really annoying to me, I would likely not use this feature as it is right now because of the annoying lines.

2) When enabling/disabling breadcrumbs, the sticky scrolling-window is not updated correctly
![9](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/c1e28ff1-afdd-4483-93f1-21e96d1e8520)

3) when doing a find/replace, the sticky lines don't update
![10](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/cb22ccc7-8a54-4b7f-a534-d8c5031c741c)

4) (optional) Line annotations aren't shown in the sticky lines
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/5872080c-840f-417b-9d4a-ee0f238d4bc0)

5) Why is there a cap on the amount of sticky lines? (I agree that >10 is likely not practical)
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/e20de6a1-d5a7-4aca-8984-89ccc6db8e59)

6) in a big file, the amount of Sticky lines should automatically decrease so that you can always still see a part of the editor. Otherwise, it can be very confusing since you cannot scroll in or correctly interact with a the editor
![11](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/5fde246f-2ea3-473a-b446-c250fff6ee19)

7) Sticky scroll sometimes "skips a line" - it's not so easy to describe this and it's not that big of a deal, but it somehow feels unnatural that in some scenarios, two lines or more are stuck at once: 
![12](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/2aed8524-d3b0-485e-b8d5-712b3c0cc683)
![13](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/40bd70fc-22ce-41d3-97df-b6fc0a862ec6)
(I know, this is quite an extreme and construed example)

8) (optional) Sticky scrolling in text files is weird:
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/217d74ee-a424-467c-ae57-b033471bea36)

9) I cannot scroll the editor text while inside of the sticky-lines area (and I would intuitively like to, especially when I have stuck 10 lines)

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/editors/text/TextEditorMessages.properties:
```diff
@@ -69,6 +69,9 @@ TextEditorDefaultsPreferencePage_showWhitespaceCharactersLinkText= (<a>configure
 TextEditorDefaultsPreferencePage_showWhitespaceCharactersDialogInvalidInput=''{0}'' is not a valid input.
 TextEditorDefaultsPreferencePage_showWhitespaceCharactersDialogTitle=Show Whitespace Characters
 TextEditorDefaultsPreferencePage_space=Space ( \u00b7 )
+TextEditorDefaultsPreferencePage_stickyScrollingEnabled=Enable sticky scrolling
```
```suggestion
TextEditorDefaultsPreferencePage_stickyScrollingEnabled=Enable stic&ky scrolling
```

k is still available for mnemonics

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyLinesProvider.java:
```diff
@@ -0,0 +1,174 @@
+ * implementation is completely based on indentation and therefore should work by default for
+ * several languages.
+ */
+public class StickyLinesProvider {
```
For the Future: how about we extract an IStickyLinesProvider and ask each editor to provide their own sticky lines provider? This fixes my comment about "normal text files" having weird behavior

Christopher-Hermann replied:
Yes, I also though about that. But I wanted to wait until we have two implementations to have a better feeling what is needed before creating a new API.

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyScrollingControl.java:
```diff
@@ -0,0 +1,482 @@
+		}
+	}
+
+	private void enusreSourceViewerLineVisible(int line) {
```
```suggestion
	private void ensureSourceViewerLineVisible(int line) {
```

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyScrollingControl.java:
```diff
@@ -0,0 +1,482 @@
+		}
+	}
+
+	public void setSettings(StickyScrollingControlSettings settings) {
```
```suggestion
	public void applySettings(StickyScrollingControlSettings settings) {
```

But that's just preference ;)

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyLinesProvider.java:
```diff
@@ -0,0 +1,174 @@
+		StyledText textWidget= sourceViewer.getTextWidget();
+		int lineCount= textWidget.getLineCount();
+
+		for (int i= startLine; i < firstBelowControl && i < lineCount; i++) {
```
Is this a safe thing to do? For larger files, this can potentially become a problem.

Scrolling with Sticky Scrolling in a huge file is already not super smooth for me:
![14](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/109998bb-6cc0-4a70-a9d6-d7e2d5e033e7)
vs
![16](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/ea5c50f9-1d82-4035-a3a0-9764c89590f7)

A possible solution is to wait until scrolling is "done" to perform this analysis - and maybe do it in a separate thread (eclipse already has enough performance-issues in the UI-thread ;_;)

I don't believe this is too big of a deal since this feature can be opted out of easily

Christopher-Hermann replied:
The for loop does only check up to the maximum amount of sticky lines. So this should not be the bottle neck. The performance problem is caused by the fast scrolling  as the sticky lines are recalculated for every moved pixel. 

Waiting until the scrolling is done is not a proper solution, I think this will look buggy / laggy.

I will try to either move the calculation in another thread or skip some calculations when the editor is scrolled really fast.

Christopher-Hermann replied:
See https://github.com/eclipse-platform/eclipse.platform.ui/pull/1949

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyLinesProvider.java:
```diff
@@ -0,0 +1,174 @@
+		return lineNumber;
+	}
+
+	private int getStartIndetation(int startFromLine, StyledText styledText) {
```
```suggestion
	private int getStartIndentation(int startFromLine, StyledText styledText) {
```

Wittmaxi commented on bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyScrollingControl.java:
```diff
@@ -0,0 +1,482 @@
+		}
+	}
+
+	private String fillLineNumberWithLeadingSpaces(int lineNumber) {
```
Isn't there a way to use an [alignment option of SWT](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fswt%2Fcustom%2FStyledText.html) when displaying the line numbers instead of padding the line number with spaces?

Christopher-Hermann replied:
I tried to implement right alignment but ended up mirroring the logic used in the line number ruler, as it also lacks specified alignment. (see [LineNumberColumn.java](https://github.com/eclipse-platform/eclipse.platform.ui/blob/385b30bb74bce1bc048584281a7b4fdedc028456/bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/source/LineNumberRulerColumn.java#L546C2-L546C11)).

Wittmaxi reviewed (approved):
The code looks good! I have added a few-nitpicks and there is one potential performance issue that I would like to hear your thoughts on :)

Christopher-Hermann commented:
Thanks for your valuable review and input. This helps a lot.

> 1. Have you looked into this change in dark mode?
> 2. When enabling/disabling breadcrumbs, the sticky scrolling-window is not updated correctly

On MacOS it's working. I need to check how I can fix and test this on windows.
![2024-06-10_17-17-30](https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/ac9681e1-3bf6-41d4-8fc8-d3d97ec3fdee)

> 4. (optional) Line annotations aren't shown in the sticky lines

You mean the blue coloring, right? This does only makes sense for the last sticky line, since all other sticky lines will be related to another source code part. Not sure if it is worth to invest here.

> 5. Why is there a cap on the amount of sticky lines? (I agree that >10 is likely not practical)

Yes, it seemed to be unnecessary to have more than 10 lines. Of course we could remove this limitation, but on the other hand this goes in the same direction as your 6. comment.

mickaelistria commented:
I'm merging this right now because the current code and feature quality are already good and it's the right time to push such experimental features in.
@Wittmaxi Can you please open dedicated issues for the point you've mentioned above that still require some improvements?

jukzi commented:
I need to get used to it, but looks promising. Just annoying that it feels like the row count is sometimes +-1 line off
Always for (empty) lines after javadoc before a method:
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/51790620/306b1c77-08d0-4ef6-94bd-58cffcdd6a92)

In this example it is unsure to which blockstart the block belongs, depends how much pixels exactly i scroll:
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/51790620/430cd101-bd02-43f5-a49a-03b06b40387a)
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/51790620/8c04f266-9526-4e0c-9be4-2925b4dbba58)
